internal/tools/aws/database_multitenant.go

```diff
@@ -0,0 +1,637 @@
+	}
+
+	if !databaseClusterInstallations.Contains(d.installationID) {
+		databaseClusterInstallations.Add(d.installationID)
+
+		err = databaseCluster.SetInstallations(databaseClusterInstallations)
+		if err != nil {
+			return errors.Wrapf(err, "failed to set installations in database cluster ID %s", *lockedCluster.cluster.DBClusterIdentifier)
+		}
+
+		err = store.UpdateDatabaseCluster(databaseCluster)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update database cluster ID %s", *lockedCluster.cluster.DBClusterIdentifier)
+		}
+	}
+
+	err = d.updateCounterTag(lockedCluster.cluster.DBClusterArn, databaseClusterInstallations.Size())
+	if err != nil {
+		return errors.Wrapf(err, "failed to set update counter tag for RDS cluster ID %s", *lockedCluster.cluster.DBClusterIdentifier)
+	}
+
+	logger.Infof("Updated counter tag multitenant RDS cluster to %d", databaseClusterInstallations.Size())
+
+	return nil
+}
+
+// Helpers
+
+type lockRDSClusterOutput struct {
+	unlock  func(log.FieldLogger)
+	cluster *rds.DBCluster
+}
+
+func (d *RDSMultitenantDatabase) lockRDSCluster(vpcID string, store model.InstallationDatabaseStoreInterface, logger log.FieldLogger) (*lockRDSClusterOutput, error) {
+	clusters, err := store.GetDatabaseClusters()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(clusters) < 1 {
+		resourceNames, err := d.client.resourceTaggingGetAllResources(gt.GetResourcesInput{
+			TagFilters: []*gt.TagFilter{
+				{
+					Key:    aws.String(trimTagPrefix(RDSMultitenantPurposeTagKey)),
+					Values: []*string{aws.String(RDSMultitenantPurposeTagValueProvisioning)},
+				},
+				{
+					Key:    aws.String(trimTagPrefix(RDSMultitenantOwnerTagKey)),
+					Values: []*string{aws.String(RDSMultitenantOwnerTagValueCloudTeam)},
+				},
+				{
+					Key:    aws.String(trimTagPrefix(DefaultRDSMultitenantVPCIDTagKey)),
+					Values: []*string{&vpcID},
+				},
+				{
+					Key: aws.String(trimTagPrefix(RDSMultitenantInstallationCounterTagKey)),
+				},
+			},
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get available RDS cluster resources")
+		}
+
+		var rdsClusterID *string
+		var installationCounter *string
+		for _, resource := range resourceNames {
+			for _, tag := range resource.Tags {
+				if *tag.Key == trimTagPrefix(RDSMultitenantInstallationCounterTagKey) && tag.Value != nil {
+					installationCounter = tag.Value
+				}
+
+				if *tag.Key == trimTagPrefix(DefaultRDSMultitenantDBClusterIDTagKey) && tag.Value != nil {
+					rdsClusterID = tag.Value
+				}
+
+				if rdsClusterID != nil && installationCounter != nil {
+					counter, err := strconv.Atoi(*installationCounter)
+					if err != nil {
+						return nil, err
+					}
+
+					if counter < DefaultRDSMultitenantDatabaseCountLimit {
+						cluster := model.DatabaseCluster{
+							ID: *rdsClusterID,
+						}
+
+						err := store.CreateDatabaseCluster(&cluster)
```

Comment:
I'm a little bit confused as to why a helper method called `lockRDSCluster` has an execution branch that may create a DatabaseCluster in the local database. I think adding some comments and breaking `lockRDSCluster` into several functions (it is quite long) would aid the clarity of this helper quite a bit.

It seems like a local representation of the RDSCluster is created if one is found in AWS but it isn't in the local database. Then the lock is stored with the new RDSCluster object in the local database. Therefore, if a new Provisioner database is used, any existing infrastructure will be considered unlocked. Is this a correct interpretation of this implementation?

Reply:
That is the correct interpretation. It's a relatively complex logic and I agree that it should be broken up in smaller helpers for clarity. My plan was to go for correctness first, get a quick review pass, add tones of testing and then make the code cleaner. However, it's great that you are pointing out all the places where it should happen so I know by a fact that it's difficult to read.